internal/datastore/crdb/tx.go

```diff
@@ -39,17 +54,19 @@ type executeTxRetryFunc func(context.Context, conn, pgx.TxOptions, transactionFn
 
 func executeWithMaxRetries(max int) executeTxRetryFunc {
 	return func(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transactionFn) (err error) {
-		return execute(ctx, conn, txOptions, fn, max)
+		return executeWithResets(ctx, conn, txOptions, fn, max)
 	}
 }
 
-// adapted from https://github.com/cockroachdb/cockroach-go
-func execute(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transactionFn, maxRetries int) (err error) {
+// executeWithResets executes transactionFn and resets the tx when ambiguous crdb errors are encountered.
+func executeWithResets(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transactionFn, maxRetries int) (err error) {
+	var retries, resets int
+	defer func() {
+		retryHistogram.Observe(float64(retries))
+		resetHistogram.Observe(float64(resets))
+	}()
+
 	var tx pgx.Tx
-	tx, err = conn.BeginTx(ctx, txOptions)
-	if err != nil {
-		return err
-	}
 	defer func() {
 		if err == nil {
 			_ = tx.Commit(ctx)
```

Comment:
What does `tx.Commit` return?

Reply:
We don't need to handle that error?

Reply:
Commit will return an error either for the tx already being close which can be ignored or to rollback. Updated to call rollback on the appropriate error.